crates/core/c8y_api/src/json_c8y.rs

```diff
@@ -286,41 +288,47 @@ pub struct C8yClearAlarm {
     pub source: Option<SourceInfo>,
 }
 
-impl From<&ThinEdgeAlarm> for C8yAlarm {
-    fn from(alarm: &ThinEdgeAlarm) -> Self {
-        let source = Self::convert_source(alarm);
-        let alarm_type = Self::convert_alarm_type(alarm);
-
-        match alarm.data.as_ref() {
-            None => C8yAlarm::Clear(C8yClearAlarm { alarm_type, source }),
-            Some(tedge_alarm_data) => C8yAlarm::Create(C8yCreateAlarm {
-                alarm_type: alarm_type.clone(),
-                source,
-                severity: C8yCreateAlarm::convert_severity(tedge_alarm_data),
-                text: C8yCreateAlarm::convert_text(tedge_alarm_data, &alarm_type),
-                time: C8yCreateAlarm::convert_time(tedge_alarm_data),
-                fragments: C8yCreateAlarm::convert_extras(tedge_alarm_data),
-            }),
+impl C8yAlarm {
+    pub fn try_from(
+        alarm: &ThinEdgeAlarm,
+        entity_store: &EntityStore,
+    ) -> Result<Self, C8yAlarmError> {
+        if let Some(entity) = entity_store.get(&alarm.source) {
+            let source = Self::convert_source(entity);
+            let alarm_type = Self::convert_alarm_type(&alarm.alarm_type);
+
+            let c8y_alarm = match alarm.data.as_ref() {
+                None => C8yAlarm::Clear(C8yClearAlarm { alarm_type, source }),
+                Some(tedge_alarm_data) => C8yAlarm::Create(C8yCreateAlarm {
+                    alarm_type: alarm_type.clone(),
+                    source,
+                    severity: C8yCreateAlarm::convert_severity(tedge_alarm_data),
+                    text: C8yCreateAlarm::convert_text(tedge_alarm_data, &alarm_type),
+                    time: C8yCreateAlarm::convert_time(tedge_alarm_data),
+                    fragments: C8yCreateAlarm::convert_extras(tedge_alarm_data),
+                }),
+            };
+            Ok(c8y_alarm)
+        } else {
+            Err(C8yAlarmError::UnsupportedExternalDeviceId(
```

Comment:
The error type is not appropriate: the source is device topic identifier not an external device id.

Reply:
Done with 816a8cb06525fbaef157ef25690ba983ed647949